### ISSUE TITLE
[Question] In SB3's PPO, how does the critic network update its weights when using separate actor and critic networks?

### ISSUE THREAD

**Issue by ytj254** [DLR-RM/stable-baselines3#1533]
### ❓ Question

Hi, I am training PPO in a custom environment where the value loss is much larger than the policy loss and dominates the entire loss function. Also, I've read some papers indicating that separate actor and critic networks would perform better. So I tried to use separate networks and tune the vf_coef to zero to eliminate the impact of the value loss. However, when I checked the source code, I did not find the relevant code for updating the critic network separately. In the PPO code, it seems that only shared actor and critic networks can be updated simultaneously. So, I would like to know if my idea of tuning vf_coef to zero makes any sense. If not, how can I deal with the different magnitude between value loss and policy loss?

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [X] If code there is, it is minimal and working
- [X] If code there is, it is formatted using the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by araffin** [comment]
Hello,
long story short, if you use SB3 > v1.8.0, the actor/critic networks are separated by default (the part that is not the feature extractor, see doc).
In that case the `vf_coef` only changes the magnitude of the critic gradient but has not influence on the actor (because they are independent).
The combined loss might be misleading, but actor and critic are not updated with the same gradients when the networks are independents.

If you put the `vf_coef` to zero, it is as if setting the value function optimizer learning rate to zero (I need to double check that statement, but I'm 90% sure ;)).

**Comment by ytj254** [comment]
Thanks for the reply. I am using SB3 2.0.0a5.
Can you please show me which line in the SB3's source code updates these two networks separately? Cause I am learning the algorithm and trying to code my own PPO agent. In the SB3's PPO code, I only found line 272 to do this work and use the combined loss to do so.
Besides, regarding the different magnitudes of actor/critic loss, can I just tune the vf_coef accordingly? Or it is better to normalize the rewards?

**Comment by araffin** [comment]
> Can you please show me which line in the SB3's source code updates these two networks separately? 

This is more related to how PyTorch/automatic differentiation works.
In short, the gradient of the actor loss with respect to the critic parameters is zero, so the actor loss won't affect the critic parameters. The same goes for the critic loss.

>  Or it is better to normalize the rewards?

See our doc, best is to normalize the reward by scaling them (we have a wrapper for that).

**Comment by ytj254** [comment]
> This is more related to how PyTorch/automatic differentiation works. In short, the gradient of the actor loss with respect to the critic parameters is zero (because the actor/critic networks are independent), so the actor loss won't affect the critic parameters. The same goes for the critic loss.

So, this one-line code "loss.backward()" can update these two separate networks simultaneously using just one combined loss function, without affecting each other. That is the same as defining two loss functions separately for actor/critic networks and using two lines of code like "pi_loss.backward()" and "vf_loss.backward()" to do the work, am I understanding that right?

> See our doc, best is to normalize the reward by scaling them (we have a wrapper for that).

I am using the "VecNormalize" function to do the normalization job, is that right?

**Comment by ytj254** [comment]
Thanks a lot!
